Tests/PromiseProgressTests.cs

```diff
@@ -0,0 +1,188 @@
+﻿using System;
+using Xunit;
+
+namespace RSG.Tests
+{
+    public class PromiseProgressTests
+    {
+        [Fact]
+        public void can_report_simple_progress()
+        {
+            var expectedStep = 0.25f;
+            var currentProgress = 0f;
+            var promise = new Promise();
+
+            promise.Progress(v =>
+            {
+                Assert.InRange(expectedStep - (v - currentProgress), -Math.E, Math.E);
+                currentProgress = v;
+            });
+
+            for (float progress = 0.25f; progress < 1f; progress += 0.25f)
+                promise.ReportProgress(progress);
+            promise.ReportProgress(1f);
+
+            Assert.Equal(1f, currentProgress);
+        }
+
+        [Fact]
+        public void can_handle_onProgress()
+        {
+            var promise = new Promise();
+            var progress = 0f;
+
+            promise.Then(null, null, v => progress = v);
+
+            promise.ReportProgress(1f);
+
+            Assert.Equal(1f, progress);
+        }
+
+        [Fact]
+        public void can_handle_chained_onProgress()
+        {
+            var promiseA = new Promise();
+            var promiseB = new Promise();
+            var progressA = 0f;
+            var progressB = 0f;
+
+            promiseA
+                .Then(() => promiseB, null, v => progressA = v)
+                .Progress(v => progressB = v)
+                .Done();
+
+            promiseA.ReportProgress(1f);
+            promiseA.Resolve();
+            promiseB.ReportProgress(2f);
+            promiseB.Resolve();
+
+            Assert.Equal(1f, progressA);
+            Assert.Equal(2f, progressB);
+        }
+
+        [Fact]
+        public void can_do_progress_weighted_average()
+        {
+            var promiseA = new Promise();
+            var promiseB = new Promise();
+            var promiseC = new Promise();
+
+            var expectedSteps = new float[] { 0.1f, 0.2f, 0.6f, 1f };
+            var currentProgress = 0f;
+            int currentStep = 0;
+
+            promiseC.
+                Progress(v =>
+                {
+                    Assert.InRange(currentStep, 0, expectedSteps.Length - 1);
```

Comment:
If the progress callback wasn't called, these asserts would not be checked. Should also have some method of ensuring that the callback has been called the correct number of times, like a counter that is incremented each time it's called and then verified at the end of the test.